### PR TITLE
CUDA extension for ExaPF

### PR DESCRIPTION
**Pull request by michel2323**

### CODE REVIEW

File: ext/ExaPFCUDAExt.jl [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+using LinearAlgebra
+using KernelAbstractions
+using SparseArrays
+
+const KA = KernelAbstractions
+const LS = ExaPF.LinearSolvers
+const PS = ExaPF.PowerSystem
+const AD = ExaPF.AutoDiff
+
+ExaPF.LS.DirectSolver(J::CUSPARSE.CuSparseMatrixCSR; options...) = ExaPF.LS.DirectSolver(nothing)
```
Comment (frapac) [inline ext/ExaPFCUDAExt.jl:22]: maybe remove `ExaPF.` here, as `LS` is already defined on line 18? Same on the following lines

File: ext/cuda_wrapper.jl [hunk 2]
```diff
@@ -31,34 +25,34 @@ function Base.unsafe_wrap(Atype::Type{CUDA.CuArray{T, 1, CUDA.Mem.DeviceBuffer}}
     unsafe_wrap(CUDA.CuArray{T, 1}, p, (dim,); own, ctx)
 end
 
-CuSparseMatrixCSR{Tv, Int32}(A::SparseMatrixCSC{Tv, Ti}) where {Tv, Ti} = CuSparseMatrixCSR(A)
+CUDA.CUSPARSE.CuSparseMatrixCSR{Tv, Int32}(A::SparseMatrixCSC{Tv, Ti}) where {Tv, Ti} = CuSparseMatrixCSR(A)
```
Comment (frapac) [inline ext/cuda_wrapper.jl:28]: I don't think we need `CUDA` here, `CUSPARSE` should be sufficient right?

File: src/LinearSolvers/preconditioners.jl [hunk 3]
```diff
@@ -233,30 +203,32 @@ function mul!(Y, C::BlockJacobiPreconditioner, B::Matrix{T}) where T
     end
 end
 
-function mul!(y, C::BlockJacobiPreconditioner, b::CuVector{T}) where T
+function mul!(y, C::BlockJacobiPreconditioner, b::AbstractVector{T}) where T
+    device = KA.get_backend(b)
```
Comment (frapac) [inline src/LinearSolvers/preconditioners.jl:207]: nice, indeed we should do that now!

File: src/Polar/second_order.jl [hunk 4]
```diff
@@ -404,13 +404,13 @@ function AutoDiff.partials!(hess::ArrowheadHessian)
     n = length(duals)
     duals_ = reshape(reinterpret(T, duals), N+1, n)
 
-    if isa(H, SparseMatrixCSC)
+    if !_iscsr(H)
```
Comment (frapac) [inline src/Polar/second_order.jl:407]: I would prefer having two functions `iscsr` and `iscsc` and write the condition as 
```julia
if iscsc(H)
    baba
elseif iscsr(H)
    abab
else 
    error("Not defined")
end
```
what do you think?

**Review by frapac — APPROVED** [review]
I think this is a net improvement to ExaPF. Thank you for working on this!

File: src/Polar/second_order.jl (hunk 4 above)
Comment (michel2323) [reply src/Polar/second_order.jl]: Done.

**Pull request merged**